### PR TITLE
ANW-1143 Optional PUI authentication

### PR DESCRIPTION
**Pull request by lorawoodford**
Adds an opt-in authentication wall for the PUI, controlled by a new `pui_require_authentication` config option (default off, so existing installs see no change in behavior).

Access is governed by a `view_pui` permission, computed live the same way `is_admin` already is. Staff members who already have PUI-viewer access are handed off transparently from their Staff UI session rather than being asked to log in a second time.

This is a global on/off gate for the whole PUI, not per-record or per-repository.  Record visibility is unaffected and still governed by the existing public-anonymous access model.

Note: I started this prior to joining the team and it comes from a fork, so the `fix_locales` failure is expected.

## Related JIRA Ticket or GitHub Issue

[ANW-1143](https://archivesspace.atlassian.net/browse/ANW-1143)

## Screenshots (if appropriate):
New PUI login:
Screenshot 2026-07-24 at 1 23 40 PM

PUI login for a user without permissions or that does not exist:
Screenshot 2026-07-24 at 1 23 59 PM

PUI logged in user:
Screenshot 2026-07-24 at 1 24 19 PM

PUI logout:
Screenshot 2026-07-24 at 1 24 27 PM

SUI make PUI viewer:
Screenshot 2026-07-24 at 1 26 49 PM

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have authority to submit this code.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

[ANW-1143]: https://archivesspace.atlassian.net/browse/ANW-1143?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: frontend/app/controllers/session_controller.rb [hunk 1]
```diff
@@ -38,6 +38,14 @@ def become_user
 
 
   def logout
+    if AppConfig[:pui_require_authentication] && AppConfig.has_key?(:public_proxy_url)
```
Comment (thimios) [inline frontend/app/controllers/session_controller.rb]: This is trying to logout the public session when the user logs out from the staff UI. Let's do that directly in the backend instead or redirect the user's browser to also perform a PUI logout.

This rails backend to rails backend request cannot clear the cookie from the user's browser.

File: backend/spec/controller_user_spec.rb [hunk 2]
```diff
@@ -243,13 +294,6 @@
     expect(last_response.status).to eq(412)
   end
 
-
```
Comment (lorawoodford) [inline backend/spec/controller_user_spec.rb:246]: Minor cleanup: This duplicates ln 65.

File: frontend/app/controllers/groups_controller.rb [hunk 3]
```diff
@@ -40,7 +40,7 @@ def update
     params[:group][:member_usernames] ||= []
 
     handle_crud(:instance => :group,
-                :model => Accession,
```
Comment (brianzelip) [inline frontend/app/controllers/groups_controller.rb:43]: @lorawoodford what was going on here such that the hardcoded `Accession` could be replaced by `JSONModel(:group)`?

File: public/app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      begin
+        with_backend_session(session[:session]) { JSONModel::HTTP.post_form('/logout') }
+      rescue StandardError => e
+        Rails.logger.error("SessionsController#logout: could not reach the backend (#{e.class}: #{e.message})")
+        Rails.logger.error("Stacktrace:\n%s" % [e.backtrace.join("\n")])
+      end
+    end
+
+    reset_session
+    redirect_to '/', notice: "Logged out successfully."
```
Comment (brianzelip) [inline public/app/controllers/sessions_controller.rb]: `notice: "Logged out successfully."`

@lorawoodford should this value should be localized?

File: public/app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -54,7 +54,7 @@
 	</div>
 
 	<section id="content" class="container-fluid mt-2 pt-2 flex-grow-1">
-		<a id="maincontent" tabindex="-1"></a>
+		<a id="maincontent" tabindex="-1" href="#maincontent"></a>
```
Comment (brianzelip) [inline public/app/views/layouts/application.html.erb]: This looks off. It's likely that the added href here should be removed, and a proper skipnav + `a#maincontent` elements should be added to the new public/app/views/layouts/login.html.erb.

Really, the new public/app/views/layouts/login.html.erb should likely be supplanted by the this application layout.

File: public/app/views/layouts/login.html.erb
Comment (brianzelip) [inline public/app/views/layouts/login.html.erb:1]: Can this file be supplanted by the application layout?

File: public/app/views/shared/_navigation.html.erb
Comment (brianzelip) [inline public/app/views/shared/_navigation.html.erb:1]: Is this dropdown consistent with others?

what about the icon used, and padding, hoverability, etc.?

File: public/app/views/shared/login.html.erb
Comment (brianzelip) [inline public/app/views/shared/login.html.erb:1]: Can we improve the presentation of this login form?

File: public/spec/controllers/application_controller_spec.rb [hunk 6]
```diff
@@ -345,3 +345,96 @@
...
+end
+
+describe WelcomeController, type: :controller do
+  context 'when pui_require_authentication is disabled' do
+    before(:each) do
+      allow(AppConfig).to receive(:[]).and_call_original
+      allow(AppConfig).to receive(:[]).with(:pui_require_authentication).and_return(false)
+    end
+
+    it 'renders the real page with no session' do
```
Comment (brianzelip) [inline public/spec/controllers/application_controller_spec.rb:360]: Maybe replace "real page" with "welcome page", since even the login page is real.

File: public/spec/controllers/application_controller_spec.rb [hunk 7]
```diff
@@ -345,3 +345,96 @@
...
+
+    it 'renders the login screen in place when there is no session' do
+      get :show
+
+      expect(response).to have_http_status(:unauthorized)
+      expect(response).to render_template('shared/login')
+      expect(response).to render_template(layout: 'layouts/login')
+    end
+
+    it 'passes through to the real page when the backend confirms a pui viewer session' do
```
Comment (brianzelip) [inline public/spec/controllers/application_controller_spec.rb:381]: Maybe replace "real page" with "welcome page", since even the login page is real.

File: public/spec/controllers/sessions_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+      end
+
+      it 'renders the shared/login view' do
+        get :show
+        expect(response).to render_template('shared/login')
+      end
+
+      it 'renders the minimal login layout, not the full application shell' do
+        get :show
+        expect(response).to render_template(layout: 'layouts/login')
```
Comment (brianzelip) [inline public/spec/controllers/sessions_controller_spec.rb:59]: Seems like this expectation could be included in the previous example, thus eliminating its example.

File: public/spec/controllers/sessions_controller_spec.rb
Comment (brianzelip) [inline public/spec/controllers/sessions_controller_spec.rb:1]: Seems like some shared_examples could be extracted from the duplicate test set up code in this file.

File: frontend/app/controllers/groups_controller.rb (hunk 3 above)
Comment (lorawoodford) [reply frontend/app/controllers/groups_controller.rb:43]: Good thing to ask about. Taking a step back: the reason this fix is in this PR at all is that an earlier version of this branch made heavy additions to this controller to add a `pui_viewer` group. Since I was already in the file for that, I fixed this along the way. That other work has since been backed out entirely (we changed how the PUI-viewer permission gets created/managed on the staff side), but I kept this fix since it's a real, independent bug.

`handle_crud`'s `:model` option is used for exactly one thing: `model.schema` gets passed into `cleanup_params_for_schema`, which walks the schema's `properties` to normalize array-shaped params before they're merged into the object via `obj.update(instance)` in the `update` action.

The old code passed `Accession` as the model — so a group edit was being cleaned against the *Accession* schema instead of the *Group* schema. `Accession`'s schema has no `member_usernames` or `grants_permissions` properties at all, so the array-fixup step was silently skipped for both fields on every group update — and those two are the only array-type fields Group has, so this wasn't a peripheral miss.

As to why this never surfaced as a visible bug: `cleanup_params_for_schema` only matters when a param arrives as a `Hash` like `name="foo[0]"`. Both fields here are submitted using a trailing `[]` convention instead (`group[grants_permissions][]` and `group[member_usernames][]`), so they're always parsed into a real `Array`, not a `Hash`. So the cleanup method never did anything for either field, and the wrong schema reference never actually changed behavior for this form. It's still always been wrong, though, so worth fixing anyway cause it's super confusing why you'd see reference to an `Accession` schema in a groups controller.

File: frontend/app/controllers/session_controller.rb [hunk 9]
```diff
@@ -38,22 +38,37 @@ def become_user
...
+      return
+    end
+
+    if params[:uri]
+      render json: check_user_access(params)
+    elsif !AppConfig[:pui_require_authentication]
       render json: { can_access: false, mode: nil }
+    elsif user_can_view_pui?
+      render json: {
+        session: session[:session],
```
Comment (thimios) [inline frontend/app/controllers/session_controller.rb]: This will hand the staff token of a potentially priviledged user to the PUI javascript running in the users browser (`authenticated.js`). 

By the way this should also be added to filter_parameters in the rails config, which is currently only `[:password]`, so that the token is not written in plaintext to the PUI's production log.

The Proposed fix would be to issue a separate PUI session instead of sharing the staff one.

The PUI reads all records and search results through the `AppConfig[:public_username]` account, not the visitor's session. A visitor's token is used for exactly two calls: GET /users/current-user and POST /logout. So the token handed to the public site can be restricted to those two, which makes a stolen one nearly worthless.

More concretely:

  - Mark PUI sessions pui_only in the session store and only allow those two requests to use such a session token.
  - Add a backend endpoint that exchanges an authenticated staff session for a PUI only one, (if the user has view_pui permission) and have check_session request a PUI session from the backend endpoint and return that token rather than the SUI session from session[:session].
  - Apply the same mechanism (use the same backend endpoint?) to create PUI specific sessions created by logging directly on the PUI.
  - Record the association between SUI session and PUI session so logout invalidates both sessions. 
  - In any case add :session to filter_parameters in both SUI and PUI rails config.

**Review by thimios — CHANGES_REQUESTED** [review]

File: frontend/app/controllers/session_controller.rb (hunk 9 above)
Comment (lorawoodford) [reply frontend/app/controllers/session_controller.rb]: @thimios This is a great solution.  Thanks for the detailed writeup. It solves the exact concern I raised in our team Slack thread (whether to move to a mint/redeem, short-lived-code exchange for the staff→PUI handoff), and I actually prefer this approach over the mint/redeem idea I'd sketched there.  This is both lighter-weight and a better fit for the codebase's existing session/permission patterns, which is what was giving me pause about my mint/redeem mechanism.

Thanks also for catching the `filter_parameters`! 

I'll start putting this together and ping you with questions if I have any.

File: public/spec/config/initializers/filter_parameter_logging_spec.rb [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+require 'spec_helper'
+require 'rails_helper'
+
+describe 'filter_parameter_logging initializer' do
+  it 'configures password and session as filtered parameters' do
+    expect(Rails.application.config.filter_parameters).to include(:password, :session)
+  end
+
+  it 'masks a session token so it is never written to the log in plaintext' do
```
Comment (thimios) [inline public/spec/config/initializers/filter_parameter_logging_spec.rb:11]: I would remove this test, I think this is testing ActiveSupport's filtering code which is out of scope.

**Review by thimios — CHANGES_REQUESTED** [review]
Found one more issue with the updated flow. If a staff user logs in two separate PUI sessions, paired_session_ids is only ever written onto the staff session, so a PUI session's /logout can follow paired_session_id up to the staff session but never back down to its siblings, so additional PUI sessions will survive the logout.

See a spec that demonstrates the issue along with a proposed solution at: https://github.com/lorawoodford/archivesspace/pull/3/changes

Also note that this avoids storing session tokens directly in the db, instead storing the digest.

I pushed a second commit to that PR that also avoids logging session tokens.

File: public/spec/config/initializers/filter_parameter_logging_spec.rb (hunk 10 above)
Comment (lorawoodford) [reply public/spec/config/initializers/filter_parameter_logging_spec.rb]: Removed ✅

**Comment by lorawoodford** [comment]
@thimios @brianzelip - This should be ready for a re-review.  Two call-outs:

1. Since fork PRs can't build the Docker images the e2e test suite needs, I verified my new pui end-2-end tests via a demo branch on the main repo: https://github.com/archivesspace/archivesspace/actions/runs/34605113795
2. In a final round of manual testing, I discovered another edge case that - while I'm calling it out here - I think is a totally acceptable trade-off given the scope of this work. If a user logs in directly on the PUI in one tab, then separately logs into the SUI in another tab, and then, finally, opens an additional PUI tab via the SUI-to-PUI handoff flow, they end up with two independent PUI sessions that aren't aware of each other.  Logging out of one doesn't affect the other. This can only happen via a fairly deliberate multi-tab, multi-login-path sequence, and the worst outcome is a confusing "I thought I logged out" moment, not any actual session/permission leak. But since I discovered it, I wanted to report it.

Anyway, ready for another look whenever you have time, and thanks for your patience with this and all the fixes.

File: public/app/controllers/sessions_controller.rb [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    parsed_body = begin
+      JSON.parse(response.body)
+    rescue JSON::ParserError
+      nil
+    end
+
+    if response.code == '200' && parsed_body
+      session[:session] = parsed_body['session']
+      session[:pui_username] = parsed_body['user']['username']
+      redirect_to '/'
```
Comment (thimios) [inline public/app/controllers/sessions_controller.rb]: should we replace '/' with `root_path` in case a proxy prefix is configured?

File: public/app/controllers/sessions_controller.rb [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        with_backend_session(session[:session]) { JSONModel::HTTP.post_form('/logout') }
+      rescue StandardError => e
+        Rails.logger.error("SessionsController#logout: could not reach the backend (#{e.class}: #{e.message})")
+        Rails.logger.error("Stacktrace:\n%s" % [e.backtrace.join("\n")])
+      end
+    end
+
+    reset_session
+    session[:skip_pui_autocheck] = true
+    redirect_to '/', notice: "Logged out successfully."
```
Comment (thimios) [inline public/app/controllers/sessions_controller.rb]: should we replace '/' with `root_path` in case a proxy prefix is configured?

File: public/app/controllers/sessions_controller.rb [hunk 13]
```diff
@@ -0,0 +1,77 @@
+class SessionsController < ApplicationController
+  skip_before_action :authenticate_user!
+  layout 'login'
+
+  def show
+    return head :forbidden unless AppConfig[:pui_require_authentication]
+    return redirect_to('/') if pui_auth_status == :ok
```
Comment (thimios) [inline public/app/controllers/sessions_controller.rb]: should we replace '/' with `root_path` in case a proxy prefix is configured?

**Review by thimios — CHANGES_REQUESTED** [review]
See also the spec in https://github.com/lorawoodford/archivesspace/pull/6 if the feature is disabled, updating a user results in silently removing their is_pui_viewer setting. I am not sure how to (or even if we should) fix that though...

**Comment by lorawoodford** [comment]
@thimios — I believe I've addressed everything from this round:

- Partially fixed the issue where the PUI-viewer flag could be silently dropped on update when its form control isn't visible, and added test coverage. (For more, see slack)
- Replaced the hardcoded root-path redirects.
- Also confirming local e2e run is still green (since it can't run directly from this fork).

**Review by thimios — APPROVED** [review]

**Review by brianzelip — APPROVED** [review]
Looks good @lorawoodford!
